extractors/src/main/scala/sqlest/extractor/Extractor.scala

```diff
@@ -132,6 +156,32 @@ case class MappedExtractor[Row, A, B](inner: Extractor[Row, A], func: A => B, un
   def emit(accumulator: inner.Accumulator) = inner.emit(accumulator).map(func)
 }
 
+case class ChoiceExtractor[Row, A, B <: D, C <: D, D](inner: Extractor[Row, A], pred: A => Boolean, left: Extractor[Row, B], right: Extractor[Row, C]) extends Extractor[Row, D] with SimpleExtractor[Row, D] with SingleRowExtractor[Row, D] {
+  type Accumulator = Option[(inner.Accumulator, Option[D])]
+
+  private def runChoice(a: A, row: Row): Option[D] = Either.cond(pred(a), runLeft(row), runRight(row)).merge
+  private def runLeft(row: Row): Option[D] = left.emit(left.initialize(row))
+  private def runRight(row: Row): Option[D] = right.emit(right.initialize(row))
+
+  def initialize(row: Row) = {
+    val innerAcc = inner.initialize(row)
+
+    inner.emit(innerAcc) map { a =>
+      (innerAcc, runChoice(a, row))
+    }
+  }
+
+  def accumulate(accumulator: Accumulator, row: Row) = {
+    for {
+      (inAcc, _) <- accumulator
+      outAcc = inner.accumulate(inAcc, row)
+      a <- inner.emit(outAcc)
+    } yield (outAcc, runChoice(a, row))
+  }
```

Comment:
Intermediate results might be discarded here. I don't think that's a problem as this is a single row extractor?